example http server

kakkun61 opened the pull request:
This example uses:

- hs-opentelemetry-api
- hs-opentelemetry-sdk
- hs-opentelemetry-propagator-datadog
- hs-opentelemetry-propagator-w3c
- hs-opentelemetry-instrumentation-wai
- hs-opentelemetry-instrumentation-http-client

iand675 committed:
Update OTLP to not use unsafe io

iand675 committed:
Move bracketError into util

iand675 committed:
Switch implementation of most context threading over to use thread-local impl

iand675 committed:
Add shutdown to example

iand675 committed:
Fix test

iand675 committed:
Check in shipped stuff

iand675 committed:
start adding some version constraints

iand675 committed:
sdk readme

iand675 committed:
me

iand675 committed:
sdk readme tweaks

iand675 committed:
Break interface slightly by exporting NewLink, bump versions

iand675 committed:
bump otlp exporter

iand675 committed:
Delete minimalform yesod example

iand675 committed:
Spec support sheet, handle more status codes in OTLP exporter

iand675 committed:
add a little bit of logging support

iand675 committed:
fiddle

iand675 committed:
Avoid using MonadTracer in instrumentation libraries

iand675 committed:
reexport makeTracer

iand675 committed:
Don't fail when getting posix user isn't possible

iand675 committed:
v0.0.3.0

iand675 committed:
janky logging test

iand675 committed:
sdk 0.0.3.2

iand675 committed:
Get onto mainline persistent

iand675 committed:
Context tweaks

Commit:
init nix

iand675 committed:
hs-opentelemetry-client 0.0.2.0 release

iand675 committed:
Forgot the changelog :(

Commit:
add hie.yaml

iand675 committed:
Fix root span cleanup with inSpan

ocharles committed:
Build with template-haskell-2.17.0

iand675 committed:
add some stack.yaml variants

iand675 committed:
Update api version

jfraudeau committed:
Fix #12: cleanup span in persistent middleware

iand675 committed:
Merge pull request #13 from jfraudeau/main

iand675 committed:
Rerelease

Commit:
cache

iand675 committed:
Merge pull request #15 from kakkun61/cache

iand675 committed:
GHC 9.2 support

Commit:
Nix for CI

Commit:
Merge branch 'main' into nix

iand675 committed:
Merge pull request #18 from kakkun61/nix

iand675 committed:
add cloudflare to repo

lf- committed:
Fix copypasted docs for alwaysOff (#19)

claymager committed:
copy into hs-opentelemetry

claymager committed:
Prettification

- Deletion of unused imports and unrelated commented-out code
- Minor refactor of inSpanM''
- run code formatter

claymager committed:
Remove more references to cloudflare

claymager committed:
Add reference to potential issues in README

claymager committed:
Include updates to generated cabal file

claymager committed:
reset tooling files

psilospore committed:
Updated docs to use makeTracer instead of deprecated getTracer

psilospore committed:
Merge pull request #1 from psilospore/psilospore-patch-1

Updated docs to use makeTracer instead of deprecated getTracer

lf- committed:
bugfix: Batch processor did not shut down cleanly, losing metrics

When the process exited, the sending of metrics to OTLP was rudely
interrupted by `cancel`. This meant that the root span would often go
missing if you instrumented some kind of command line application with
the root span created at process creation.

This patch changes the shutdown logic to first ask the exporter to send
what it has and shut down, then forcibly kill it if that takes too long.

lf- committed:
Add hspec instrumentation

It would be useful to have instrumentation of test suites. My
original use case for this is to validate a change in Persistent
instrumentation, for which hitting it really hard with the Mercury test
suite seems like a reasonable strategy.

lf- committed:
Add Show instances

I needed these for debugging, but they are generally useful.

There's an unfortunate amount of trailing whitespace deletion in this
diff. Sorry :(

parsonsmatt committed:
Support GHC 9.4

iand675 committed:
Merge pull request #25 from MercuryTechnologies/show-instances

iand675 committed:
Merge pull request #27 from parsonsmatt/matt/support-ghc-94

iand675 committed:
Merge pull request #22 from psilospore/main

iand675 committed:
Merge pull request #23 from MercuryTechnologies/batch-processor-shutdown

iand675 committed:
Merge pull request #21 from BlueChipFinancial/utils-exceptions

iand675 committed:
version bumps

lf- committed:
Add an example and docs

iand675 committed:
Merge pull request #24 from MercuryTechnologies/hspec-instrumentation

iand675 committed:
Add Jade as signficant contributor

lf- committed:
Fix accidental italics in docs

lf- committed:
Merge pull request #30 from MercuryTechnologies/fix-italics

Fix accidental italics in docs

kakkun61 committed:
format action

kakkun61 committed:
add fourmolu

kakkun61 committed:
fourmolu without nix

Nixpkgs release-22.05 channel's fourmolu is too old to parse some codes.

kakkun61 committed:
fourmolu parsable

kakkun61 committed:
format

kakkun61 committed:
proto-lens 0.7.1.2 already supports GHC 9.4

https://github.com/google/proto-lens/pull/440

kakkun61 committed:
fix stack compilation error

kakkun61 committed:
fix github action

kakkun61 committed:
update cabal files

kakkun61 committed:
config nix path

iand675 committed:
Merge pull request #31 from herp-inc/fix-compilation-error

kakkun61 committed:
add c libs

kakkun61 committed:
CI for cabal

kakkun61 committed:
cabal under nix

kakkun61 committed:
update hie.yaml

kakkun61 committed:
hpack

kakkun61 committed:
CI runs also on PR

kakkun61 committed:
Merge pull request #34 from herp-inc/ci-pr

CI runs also on PR

kakkun61 committed:
Merge branch 'main' into fix-compilation-error-cabal

kakkun61 committed:
cabal update

kakkun61 committed:
nix-shell

kakkun61 committed:
Merge pull request #33 from herp-inc/fix-compilation-error-cabal

Fix compilation error for cabal users

kakkun61 committed:
phony

kakkun61 committed:
indent-wheres: true

kakkun61 committed:
format

iand675 committed:
Merge pull request #35 from herp-inc/format

lf- committed:
Fix typo in docs

lf- committed:
Merge pull request #36 from MercuryTechnologies/fix-typo

Fix typo in docs

Commit:
add instances for monad-tracer

Commit:
StateVar and Only need libffi

Commit:
libffi for stack

parsonsmatt committed:
Fix batch shutdown deadlock

parsonsmatt committed:
ahh

parsonsmatt committed:
space

parsonsmatt committed:
manually fix formatting

parsonsmatt committed:
wow

parsonsmatt committed:
Synchronously cancel the worker

parsonsmatt committed:
Flush queue on shutdown

parsonsmatt committed:
error return refactor

iand675 committed:
Merge pull request #42 from parsonsmatt/mattp/fix-batch-shutdown

Fix batch shutdown deadlock

iand675 committed:
Release SDK and OTLP exporter shutdown fixes

lf- committed:
Add a vendor-honeycomb package with trace link stuff (#39)

* wip vendor honeycomb support

* Add a little testing

* Remove monad environment stuff

* Fix timestamp calculation

* Expose a super simple API

* correct test

* New approach using the Context vault to deal with the target

* format

* format ii

lf- committed:
sdk: Crash if -threaded is unset (#46)

* sdk: Crash if -threaded is unset

registerDelay will throw an exception, which we will silently swallow, if we
don't preemptively just blow up the program. It's never correct to start up
without a threaded runtime.

* Add a note

Commit:
overlappable

Commit:
update CI

Commit:
format with fourmolu 0.10.1.0

Commit:
freeze

Commit:
init a Datadog propagator

Commit:
Merge pull request #48 from herp-inc/update-ci

Update CI

Commit:
init persistent-mysql

Commit:
Merge remote-tracking branch 'upstream/main' into datadog-propergator

Commit:
format

Commit:
hide constructors of IDs

https://github.com/iand675/hs-opentelemetry/pull/49#discussion_r1070138527

iand675 committed:
Merge pull request #37 from herp-inc/monad-tracer-instances

add instances for monad-tracer

Commit:
add a datadog propagator to README

iand675 committed:
Merge pull request #49 from herp-inc/datadog-propergator

Datadog propagator

Commit:
add persistent-mysql to README

Commit:
fix GHC version tag about 8.10

kakkun61 committed:
persistent-mysql attributes

kakkun61 committed:
Workaround

https://github.com/haskell/actions/issues/187

kakkun61 committed:
fix CI name

kakkun61 committed:
http server example

This example uses:

- hs-opentelemetry-api
- hs-opentelemetry-sdk
- hs-opentelemetry-propagator-datadog
- hs-opentelemetry-propagator-w3c
- hs-opentelemetry-instrumentation-wai
- hs-opentelemetry-instrumentation-http-client

9999years committed:
Update instrumentation-hspec for GHC 9.6

9999years committed:
Import `join` from `Control.Monad`

Reexports removed in mtl-2.3/transformers-0.6

9999years committed:
Fix memory leak

Co-authored-by: Ian-Woo Kim @wavewave

iand675 committed:
Merge pull request #53 from herp-inc/workaround-ghcup

Workaround for ghcup issue

solomon-b committed:
Adds B3 Propagator Library

Commit:
Merge pull request #63 from 9999years/honeycomb

Fix `hs-opentelemetry-vendor-honeycomb` for mtl-2.3/transformers-0.6

solomon-b committed:
Apply Fourmolu Formatting

solomon-b committed:
Add b3 propagator to stack.yaml

Commit:
Merge pull request #62 from 9999years/instrumentation-hspec-ghc96

Update instrumentation-hspec for mtl-2.3/GHC-9.6

solomon-b committed:
Remove unused 'case-insensitive' depenednecy from b3 propagator

Commit:
Merge pull request #66 from hasura/feature/adds-b3-propagator

Adds B3 Propagator Library

kakkun61 committed:
more familiar and better perf. "show" instance

Commit:
refresh CI cache

Commit:
refresh CI cache

Commit:
show emptyAppendOnlyBoundedCollection

Commit:
Merge branch 'refresh-ci-cache' into better-show

Commit:
format

Commit:
fix compilation errors

Commit:
use bang patterns instead of seq

9999years committed:
Merge pull request #1 from herp-inc/api-fix-memory-leak

Fix compilation errors etc.

iand675 committed:
Merge pull request #55 from herp-inc/fix-ci-name

fix a CI step's name

iand675 committed:
Merge pull request #67 from herp-inc/better-show

more familiar and better performance "show" instance

iand675 committed:
Merge pull request #64 from 9999years/api-fix-memory-leak

Fix memory leak

iand675 committed:
Work on getting the nix setup usable for building & testing the codebase in CI

iand675 committed:
Add linkFarm default target

iand675 committed:
flake tweaks

iand675 committed:
test

iand675 committed:
fiddle

iand675 committed:
fiddle

iand675 committed:
maintenance

iand675 committed:
maintenance

iand675 committed:
regen cabal files

iand675 committed:
fix flake support for GHC 9.4

iand675 committed:
try flake in GHA

iand675 committed:
quotes

iand675 committed:
morale must improve

iand675 committed:
this should probably definitely work

iand675 committed:
fiddle

iand675 committed:
fiddle

iand675 committed:
this is promising

iand675 committed:
cabal

iand675 committed:
remove constraints from cabal.project

iand675 committed:
Debugging

iand675 committed:
Debugging

iand675 committed:
Don't attach empty contexts to WAI threads when finished

iand675 committed:
bump thread-utils-context

iand675 committed:
bip

iand675 committed:
not sure if post qualified works in GHC 8.x

iand675 committed:
Small documentation improvements

iand675 committed:
Cabal file wrangling

iand675 committed:
Version bumps

iand675 committed:
Add vendors/honeycomb to cabal project

iand675 committed:
Release SDK with b3 support

iand675 committed:
Fix compilation error

JonathanLorimer committed:
Fix README.md code example

iand675 committed:
Merge pull request #72 from JonathanLorimer/JonathanLorimer-patch-1

Fix README.md code example

iand675 committed:
Mercury work

iand675 committed:
Fix build?

iand675 committed:
fiddle

parsonsmatt committed:
Add instrumentSpec function

parsonsmatt committed:
format

iand675 committed:
Merge pull request #81 from parsonsmatt/mattp/hspec-describe-context

Add instrumentSpec function

iand675 committed:
Merge remote-tracking branch 'origin/main' into improved-dux

iand675 committed:
fold in hspec changes and fix the examples

chrismanning committed:
Add missing import

chrismanning committed:
Decode percent-encoded baggage headers

chrismanning committed:
Formatting

chrismanning committed:
Only import word64ToWord# for base-4.17.0 or newer

jkeuhlen committed:
Implement maxExportBatchSize in batch processor

Commit:
Merge pull request #84 from chrismanning/decode-percent-encoded-baggage

Decode percent-encoded baggage headers

chrismanning committed:
Formatting

chrismanning committed:
Formatting

Commit:
Merge pull request #83 from chrismanning/main

Add missing import

iand675 committed:
Merge pull request #85 from jkeuhlen/ImplmentMaxExportBatchSize

Implement max export batch size

Commit:
newOpenTelemetryWaiMiddleware' is now not IO action

Commit:
update change log

Commit:
Merge pull request #86 from iand675/wai-no-io

newOpenTelemetryWaiMiddleware' is now not an IO action

Commit:
Merge pull request #75 from herp-inc/stack-ghc-8.10

fix GHC version tag about 8.10

Commit:
make Attributes be HashMap

Commit:
fourmolu 0.13.1.0

Commit:
change log and version up

Commit:
Merge branch 'main' into attributes-hashmap

iand675 committed:
Merge pull request #88 from herp-inc/attributes-hashmap

Use maps instead of lists of pairs as attributes

iand675 commented on examples/http-server/main.hs:
Small nitpick: `main.hs` should probably be `Main.hs` instead.

Commit:
Merge pull request #87 from herp-inc/update-fourmolu

Use fourmolu 0.13.1.0

Commit:
Merge remote-tracking branch 'upstream/main' into persistent-mysql

Commit:
fix attributes type

Commit:
Merge pull request #50 from herp-inc/persistent-mysql

persistent-mysql instrumentation

iand675 committed:
back to stack

iand675 committed:
fiddle

iand675 committed:
fiddle

iand675 committed:
newer nix version

iand675 committed:
oodles of formatting

iand675 committed:
fourmolu action moved to haskell-actions

iand675 committed:
fiddle

iand675 committed:
Fix compilation on GHC 8.12

iand675 committed:
get unsafeCoerce# from GHC.Exts

iand675 committed:
format

iand675 committed:
Merge remote-tracking branch 'origin/main'

iand675 committed:
tag releases

parsonsmatt committed:
Swallow localhost errors

parsonsmatt committed:
lol

iand675 committed:
Merge pull request #89 from parsonsmatt/mattp/swallow-localhost-errors

Swallow localhost errors

Qqwy committed:
Fix git conflict in hs-opentelemetry-utils-exceptions.cabal

There was a committed git conflict in utils/exceptions/hs-opentelemetry-utils-exceptions.cabal

This commit fixes it.

iand675 committed:
Merge pull request #107 from Qqwy/Qqwy-patch-1

Fix git conflict in hs-opentelemetry-utils-exceptions.cabal

evanlauer1 committed:
Update HTTP Semantic Conventions to Stable Versions (#118)

* Work in Progress: Request and Response Attributes updated to match the latest stable semantics if the environment variable OTEL_SEMCONV_STABILITY_OPT_IN is "http"

* Added Stable attributes to instrumentResponse

* Refactored to fit new httpTracerProvider return type

* Fixed missing "do" notation and removed unnecessary imports

* Fix typo and update comment

* Found unchanged http name

* Found outdated http attribute names

* Fixed cradle so HLS works

* Added Settings File for application-wide environment variable configuration

* Moved semConvStabilityOptIn and related types to OpenTelemetry.Settings

* Changed code using HttpTracerProvider to work with changed return type

* Fixed exports

* Replaced general Settings.hs with specific SemConvStability.hs

* Removed comment because the problem it referenced is solved

* removed unused language pragma

* fixed imports

* Replaced outdated http conventions based on the value of OTEL_SEMCONV_STABILITY_OPT_IN

* changed line about generation version

* added unordered-containers to dependencies

* Moved semConvStabilityOptIn so that it is in scope

* Updated dependencies for testing

* WIP testing for http-client

* Generated by newer version

* Removed WIP testing

* Updated getSemConvStabilityOptIn  so it parses environment variable as a comma-separated list

* derives Show  and Eq for testing purposes

* Added tests for SemConvStabilityOptIn to make sure environment variable is read correctly

* Updated Haddock to include blurb about the new HTTP semantic conventions

* Removed hspec from dependencies

* Auto-generated change. Removed hspec from build-depends

* Added newline to end of file

* Split off parsing logic from getSemConvStabilityOptIn to parseSemConvStabilityOptIn

* Added blank line for readability

* Changed pure to Just to improve readability and contrast with Nothing

* Rewrote tests to eliminate boilerplate and increase readability

* Refactored and Renamed SemanticsConfig.hs to future proof

* Refactored SemanticsConfig so options are Enums for pattern matching instead of booleans

* Updated SemanticsConfig tests. NOTE: does not currently pass memoization tests.

* Updated to match changes to SemanticsConfig.hs

* Fix infinite loop issue

Co-authored-by: Ian Duncan <ian@iankduncan.com>

* changed from . to $

* Added haddocks

* Formatting fixes

---------

Co-authored-by: Ian Duncan <ian@iankduncan.com>

Gabriella439 committed:
Export `FlushResult` (#96)

The return value of `forceFlushTracerProvider` is `FlushResult`, but
that type isn't exported so it can't be used, and this change fixes
that.

pbrisbin committed:
Add Datadog as a Known Propagator (#117)

* Don't depend on sdk from propagator-datadog

This meant adding ourselves to `knownPropagators` would be circular. We
can easily grab this type from the `api` package instead.

* Add datadog as a known propagator

iand675 committed:
CI overhaul (#120)

* test

* Add some more GHC versions

* Add pre-commit checks and some other goodies

* missed some hooks there

* Exclude shell builds for now

* GHC 8.10 and 9.0 aren't in nixpkgs anymore?

* Get particular about build packages

* Don't torture garnix with too many packages

* fix glob

* Hardcode packages since globs are wimpy

* pre-commit checks

* turn off old checks

* Get checks to agree with each other

iand675 committed:
Add basic CONTRIBUTING.md doc

evanlauer1 committed:
Change OTLP generation strategy from manual to auto-generation (#122)

* Experimenting with protoc autogen

* Got first pass at auto-generated files working

* Auto-generated files

* Moved proto submodule into otlp

* Fixed custom setup to work with Cabal using defaultMain

* added field "autogen-modules"

* changed verbatim generated-exposed-modules

* Formatting

* Added documentation for auto-generation

* Further updated auto-generation instructions

* tweak

* experimental CI fix

* update cabal files

---------

Co-authored-by: Ian Duncan <ian@iankduncan.com>

evanlauer1 committed:
Rename `Log` to `LogRecord` and update data type to match logs data model spec (#125)

* Created `LogAttributes` where `AnyValue` is a superset of `Attribute`

* Renamed `Log` to `LogRecord` and updated fields to match current Log Data Model spec

evanlauer1 committed:
Created emitLogRecord that takes a Logger as an argument. Refactored types to a separate module. stab-267 (#126)

* Created emitLogRecord that takes a Logger as an argument. Refactored types to a separate module

* Deleted legacy implementations of SeverityNumber

* Updated export list

evanlauer1 committed:
Added instrumentationScope to Logger and LogRecord (#127)

* Added instrumentationScope to Logger and LogRecord

* Removed Hashable instance from InstrumentationLibrary and Attributes

* Added Hashable instance back to InstrumentationLibrary

* Updated comment for instrumentationLibrary

* Further updated comment

evanlauer1 committed:
Add LoggerProvider to create Loggers and allow registering a global LoggerProvider (#128)

* add-loggerProvider-to-create-loggers

* Added documentation

* Removed `Maybe` from Resources field because `emptyResources` can be used instead

* Removed MonadIO constraint from `createLoggerProvider`

evanlauer1 committed:
Refactored attribute limits to be controlled by the LoggerProvider (#129)

michaelpj committed:
Fix callerAttributes in the presence of frozen call stacks (#137)

Freezing the call stack is common when writing your own utility wrappers
around e.g. `withSpan`. In this case you want to freeze the call stack
before calling your utility, so that `withSpan` and friends see the
"application" function as the top of the call stack.

However, the functions for getting information from call stacks
just generally didn't work right in the presence of
frozen call stacks. They assumed that the call stack was
unfrozen, and hence that they could look up a fixed number of steps to
find the caller info.

We can't in general tell directly if the call stack is frozen, but we
use a pretty decent heuristic here: is the top element exactly our own
function? If so, we probably have an unfrozen call stack. Otherwise we
might have a frozen call stack and we have to try and do something else.

iand675 committed:
Otel overlay (#138)

* testing

* tweak

evanlauer1 committed:
Added Logging when creating spans or emitting logs drops attributes. Updated Severity Number (#133)

* Added Logging to emitLogRecord and createSpanWithoutCallStack when dropping attributes. Updated Severity Number

* Added comments warning users not to use emitLogRecord

* Added further documentation to warn against using emitOTelLogRecord

* Explicitly handle Unknown case in toShortName

* Changed Err to Error in SeverityNumber

* Moved Logging.Core to Internal.Logging.Core to hide functions

* Revert "Moved Logging.Core to Internal.Logging.Core to hide functions"

This reverts commit 32c5e58978aef92d94bc9042a471340daafd6002.

* Moved Logging.Core to Internal.Logging.Core

* Re-added Logging.Core with re-exported functions from Internal.Logging.Core

evanlauer1 committed:
Refactored LogRecords to be mutable so that attributes can be modified after creation (#131)

* Refactored LogRecords to be mutable so that attributes can be modified after creation

* Updated test suite

* Added ReadableLogRecord and ReadWriteLogRecord interfaces

* Added typeclasses for ReadableLogRecords and ReadWriteLogRecords

* Replaced type variable in LogRecord with AnyValue

* Changed Readable and ReadWriteLogRecord to be concrete types

* Changed mkReadableLogRecord to take in a ReadWriteLogRecord

* Added NullValue to AnyValue and removed body parameter from emptyLogRecordArguments

evanlauer1 committed:
Added LogProcessors and implemented shutdown and forceFlush for LoggerProviders (#132)

* Renamed LogProcessor to LogRecordProcessor

* Added `noOpLoggerProvider` and removed `MonadIO` constraint from `createLoggerProvider`

* Renamed `LogProcessor` to `LogRecordProcessor`

* Call processors' onEmit when a log is emitted

evanlauer1 committed:
Renamed Processor to SpanProcessor, Exporter to SpanExporter, and Logging to Logs (#134)

* Renamed Processor to SpanProcessor, Exporter to SpanExporter, and Logging to Logs

* Updated .cabal files

* Updated hie.yaml

* Updated nix haskell packages

* Renamed span-exporter modules to exporter

* Continued module restructuring

* Finished module restructuring

* Renamed SpanProcessor to Processor.Span

* Removed LogRecord from Legacy Processor and Exporter modules

* Added pattern Synonyms for Exporters and Processors

* Added deprecated warnings to Exporters and Processors

* Fixed nix haskell-packages

* Fixed typo in Zipkin exporter

Commit:
add protobuf nix dependency to stack.yaml (#140)

kakkun61 commented:
@iand675 I resolved this conflict.
